presto-main/src/main/java/com/facebook/presto/sql/planner/GroupingOperationRewriter.java

```diff
@@ -0,0 +1,114 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.sql.planner;
+
+import com.facebook.presto.metadata.Metadata;
+import com.facebook.presto.sql.analyzer.Analysis;
+import com.facebook.presto.sql.analyzer.FieldId;
+import com.facebook.presto.sql.analyzer.TypeSignatureProvider;
+import com.facebook.presto.sql.tree.ArrayConstructor;
+import com.facebook.presto.sql.tree.Cast;
+import com.facebook.presto.sql.tree.Expression;
+import com.facebook.presto.sql.tree.FunctionCall;
+import com.facebook.presto.sql.tree.GroupingOperation;
+import com.facebook.presto.sql.tree.LongLiteral;
+import com.facebook.presto.sql.tree.QualifiedName;
+import com.facebook.presto.sql.tree.QuerySpecification;
+import com.facebook.presto.sql.tree.SimpleGroupBy;
+import com.facebook.presto.type.ListLiteralType;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import static com.facebook.presto.operator.scalar.GroupingOperationFunction.GROUPING;
+import static com.facebook.presto.spi.type.BigintType.BIGINT;
+import static com.facebook.presto.sql.analyzer.ExpressionAnalyzer.resolveFunction;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+
+public final class GroupingOperationRewriter
+{
+    private GroupingOperationRewriter() {}
+
+    public static Expression rewriteGroupingOperation(GroupingOperation expression, QuerySpecification queryNode, Analysis analysis, Metadata metadata, Optional<Symbol> groupIdSymbol)
+    {
+        requireNonNull(queryNode, "node is null");
+        requireNonNull(analysis, "analysis is null");
+        requireNonNull(metadata, "metadata is null");
+        requireNonNull(groupIdSymbol, "groupIdSymbol is null");
+
+        checkState(queryNode.getGroupBy().isPresent(), "GroupBy node must be present");
+
+        // No GroupIdNode and a GROUPING() operation imply a single grouping, which
+        // means that any columns specified as arguments to GROUPING() will be included
+        // in the group and none of them will be aggregated over. Hence, re-write the
+        // GroupingOperation to a constant literal of 0.
+        // See SQL:2011:4.16.2 and SQL:2011:6.9.10.
+        boolean isSimpleGroupByPresent = queryNode.getGroupBy().get().getGroupingElements().stream().anyMatch(SimpleGroupBy.class::isInstance);
+        if (isSimpleGroupByPresent) {
+            return new LongLiteral("0");
+        }
+        else {
+            checkState(groupIdSymbol.isPresent(), "groupId symbol is missing");
+
+            Map<Expression, FieldId> columnReferenceFields = analysis.getColumnReferenceFields();
```

Comment:
So I solved the whole equality and identity problem. I was being an idiot and inside of `GroupingOperation` I was storing the arguments as `QualifiedName` and every time I wanted the arguments I would convert them to either an `Identifier` or `DereferenceExpression` using `DereferenceExpression::from(QualifiedName)`. That meant I would get new instances every time and they would never get analyzed which in turn meant that they would never be added to the `columnReferences`. So now I am storing the grouping arguments in their converted format and I do that conversion in the constructor.

This means I can used the `FieldId` instead of using `indexOf`, which I think is much more elegant. I re-wrote the stream as you suggest. I also added the below test:
```
SELECT grouping(a) FROM (VALUES ('h', 'j', 11), ('k', 'l', 7)) AS t (a, b, c) GROUP BY GROUPING SETS (a,c), c*2
```

@sopel39 I don't understand the difference and purpose between the two tests you suggest? The first aliases the table and the second doesn't but one of the columns refers to a non-existent alias? Can you clarify?